response.go

```diff
@@ -0,0 +1,248 @@
+}
+
+// UnmarshalBinary decodes the response from binary form.
+// This method implements encoding.BinaryUnmarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryUnmarshaler
+func (r *Response) UnmarshalBinary(data []byte) error {
+	_, err := asn1.Unmarshal(data, r)
+	return err
+}
+
+// ValidateStatus validates the response.Status
+//
+// Reference: RFC 3161 2.4.2
+func (r *Response) ValidateStatus() error {
+	if r.Status.Status != pki.StatusGranted && r.Status.Status != pki.StatusGrantedWithMods {
+		failureInfo, err := r.Status.ParseFailInfo()
+		if err != nil {
+			return &InvalidResponseError{Msg: fmt.Sprintf("invalid response with status code %d: %s", r.Status.Status, r.Status.Status.String())}
+		}
+		return &InvalidResponseError{Msg: fmt.Sprintf("invalid response with status code %d: %s. Failure info: %s", r.Status.Status, r.Status.Status.String(), failureInfo)}
+	}
+	return nil
+}
+
+// SignedToken returns the timestamp token with signatures.
+// Callers should invoke Verify to verify the content before comsumption.
+func (r *Response) SignedToken() (*SignedToken, error) {
+	if err := r.ValidateStatus(); err != nil {
+		return nil, err
+	}
+	return ParseSignedToken(context.Background(), r.TimeStampToken.FullBytes)
+}
+
+// Validate checks if resp is a successful timestamp response against
+// its corresponding request based on RFC 3161.
+// It is used when a timestamp requestor receives the response from TSA.
+func (resp *Response) Validate(req *Request) error {
+	if req == nil {
+		return &InvalidResponseError{Msg: "missing corresponding request"}
+	}
+	if resp == nil {
+		return &InvalidResponseError{Msg: "response cannot be nil"}
+	}
+	if err := resp.ValidateStatus(); err != nil {
+		return err
+	}
+	token, err := resp.SignedToken()
+	if err != nil {
+		return &InvalidResponseError{Detail: err}
+	}
+	info, err := token.Info()
+	if err != nil {
+		return &InvalidResponseError{Detail: err}
+	}
+	if info.Version != 1 {
+		return &InvalidResponseError{Msg: fmt.Sprintf("timestamp token info version must be 1, but got %d", info.Version)}
+	}
+	// check policy
+	if req.ReqPolicy != nil && !req.ReqPolicy.Equal(info.Policy) {
+		return &InvalidResponseError{Msg: fmt.Sprintf("policy in response %v does not match policy in request %v", info.Policy, req.ReqPolicy)}
+	}
+	// check MessageImprint
+	if !info.MessageImprint.HashAlgorithm.Algorithm.Equal(req.MessageImprint.HashAlgorithm.Algorithm) ||
+		!bytes.Equal(info.MessageImprint.HashedMessage, req.MessageImprint.HashedMessage) {
+		return &InvalidResponseError{Msg: fmt.Sprintf("message imprint in response %+v does not match with request %+v", info.MessageImprint, req.MessageImprint)}
+	}
+	// check nonce
+	if req.Nonce != nil {
+		responseNonce := info.Nonce
+		if responseNonce == nil || responseNonce.Cmp(req.Nonce) != 0 {
+			return &InvalidResponseError{Msg: fmt.Sprintf("nonce in response %s does not match nonce in request %s", responseNonce, req.Nonce)}
+		}
+	}
+	// check certReq
```

Comment:
We are missing following validations 
```
   The policy field MUST indicate the TSA's policy under which the
   response was produced.  If a similar field was present in the
   TimeStampReq, then it MUST have the same value, otherwise an error
   (unacceptedPolicy) MUST be returned.  This policy MAY include the
   following types of information (although this list is certainly not
   exhaustive):
```

```
   The messageImprint MUST have the same value as the similar field in
   TimeStampReq, provided that the size of the hash value matches the
   expected size of the hash algorithm identified in hashAlgorithm.
```

```
 The serialNumber field is an integer assigned by the TSA to each
   TimeStampToken.  <-- validate presence
```

```
validate presense of `genTime` in UTC and has precision of atleast seconds.
```

Reply:
The 1st and 2nd checks are already included in the same function.
Added 3rd and 4th checks.